packages/grid/x-data-grid/src/hooks/features/filter/gridFilterUtils.ts

```diff
@@ -295,15 +341,18 @@ export const buildAggregatedFilterApplier = (
   });
 };
 
+const filterModelItems = defaultMemoize(
```

Comment:
Instead of memoization, maybe do something like this?
```js
// Not sure about the name though
class FilterResultsMatcher {
  constructor({ apiRef filterModel }) {
    this.filterModel = filterModel;
    this.cleanedFilterItems = filterModelItems(apiRef, filterModel.items);
  }

  passFilterLogic(
    allFilterItemResults: (null | GridFilterItemResult)[],
    allQuickFilterResults: (null | GridQuickFilterValueResult)[],
  ) {
    return passFilterLogic(
      allFilterItemResults,
      allQuickFilterResults,
      this.filterModel,
      this.cleanedFilterItems,
    );
  }
}
```

Then we can use it like this:
```diff
// useGridFilter.tsx

const filteredRowsLookup: Record<GridRowId, boolean> = {};
+const filterResultsMatcher = new FilterResultsMatcher({ apiRef, filterModel: params.filterModel });
for (let i = 0; i < rowIds.length; i += 1) {
  const rowId = rowIds[i];
  let isRowPassing;
  if (typeof rowId === 'string' && rowId.startsWith('auto-generated-group-footer')) {
    isRowPassing = true;
  } else {
    const { passingFilterItems, passingQuickFilterValues } =
      params.isRowMatchingFilters(rowId);
-    isRowPassing = passFilterLogic(
+    isRowPassing = filterResultsMatcher.passFilterLogic(
      [passingFilterItems],
      [passingQuickFilterValues],
-      params.filterModel,
-      apiRef,
    );
  }
```

Reply:
Pretty sure it should work even for multi-instance, the goal is just to avoid repeating the operation in the filtering loop, and the filtering loop is synchronous. It can indeed be something else than a memoization, initially I tried adding a `context` param to `passFilterLogic` to keep re-usable data around, but the memoization required less changes. Memoization is probable the wrong approach from a conceptual POV, and the one your proposed or the `context` param are more understandeable in this case.

Reply:
So I've got a commit for this stashed, I must say I didn't prefer it because it adds more noise. Lmk what you think, I'll apply it if you really prefer this approach.

Reply:
Indeed, it shouldn't be a problem for synchronous loops since it's only used to make subsequent iterations faster 👍
Let's keep the top-level memo for simplicity